addons/onboarding/models/onboarding_step.py

```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models
+
+
+class OnboardingStep(models.Model):
+    _name = 'onboarding.onboarding.step'
+    _description = 'Onboarding Step'
+
+    name = fields.Char('Name of the onboarding step', translate=True)
+    onboarding_id = fields.Many2one('onboarding.onboarding', required=True, ondelete='cascade')
+
+    is_per_company = fields.Boolean(related='onboarding_id.is_per_company', string='Should be done per company?')
+
+    panel_step_title = fields.Char('Title', translate=True)
+    panel_step_description = fields.Char('Description', translate=True)
+    panel_step_button_text = fields.Char('Button text', default="Let's do it", translate=True)
+    panel_step_done_icon = fields.Char('Font Awesome Icon when completed', default='fa-star')
+    panel_step_done_text = fields.Char(
+        'Text to show when step is completed', default='Step Completed! - Click to review', translate=True)
+    panel_step_open_action_name = fields.Char('Action to execute when opening the step', required=True)
+
+    current_progress_id = fields.One2many(
+        'onboarding.progress.step', 'step_id', string='Step State Record',
+        compute='_compute_current_progress_id')
+    progress_ids = fields.One2many('onboarding.progress.step', 'step_id', 'State records')
+
+    _sql_constraints = [
+        ('name_onboarding_uniq',
+         'unique (name,onboarding_id)',
+         'There cannot be multiple records of the same onboarding step name.')
+    ]
+
+    @api.depends_context('company')
+    @api.depends('progress_ids', 'is_per_company')
+    def _compute_current_progress_id(self):
+        per_company = self.filtered_domain([('is_per_company', '=', True)])
+        existing_progress_steps = self.env['onboarding.progress.step'].search([
+            '|',
+            ('step_id', 'in', (self - per_company).ids),
+            '&', ('step_id', 'in', per_company.ids),
+                 ('company_id', '=', self.env.company.id),
+        ])
+
+        for step in self:
+            if step in existing_progress_steps.step_id:
+                step.current_progress_id = existing_progress_steps.filtered(
+                    lambda progress_step: (progress_step.step_id == step
+                                           and (progress_step.company_id.id in {False, self.env.company.id})))
+            else:
+                step.current_progress_id = False
+
+    def create(self, values):
+        steps = super().create(values)
+        # Create new ProgressSteps records for existing onboarding progress records only
+        steps.create_progress_records()
+        return steps
+
+    def consolidate_just_done(self):
+        self.current_progress_id.consolidate_just_done()
+
+    def set_done(self):
+        self.current_progress_id.set_done()
```

Comment:
Could we have a single "mark_done" that takes an argument to only update if the status is "just_done"?
If you need to change the JS, just keep it as is.